option optimize_for = LITE_RUNTIME;
package pb;
import public "common.proto";


// USER MESSAGE STRUCTURE
message MsgInventoryRequest {
  oneof action{
    Add add = 2;
    Remove remove = 5;
    Modify modify = 6;
    Get get = 7;
    AddShelf addShelf = 15;
//    RmShelf rmShelf = 16;
//    MoveShelf mvShelf = 17;
  };

// ACTION MESSAGES /////////////////////
  message Add{
    required string name = 1;
    optional string description = 2;
    repeated InventoryShelf shelfs = 3;
  }

  message Remove{
    required int32 uid = 1;
    optional bool removeItems = 2;
  }

  message Modify{
    required int32 uid = 1;
    optional string newName = 2;
    optional string newDesc = 3;
    optional Acl newAcl = 4;
  }

  message Get {
    message Where{
        optional int32 user_id = 1; // will return all user inventories
        optional int32 inventory_id = 2; // will return only one inventory
        optional int32 group_id = 3;
    }
    optional bool id = 1;
    optional bool name = 2;
    optional bool description = 3;
    optional bool acl = 4; // acl contains ID
    optional bool shelfs = 5;
    required Where where = 6;
  }

  message  AddShelf {
    required int32 inventoryId = 1;
    required InventoryShelf shelf = 2;
  }
}

message MsgInventoryResponse {
    enum Error {
      No_Error = 0;
      NameToLong = 1;
      NameEmpty = 2;
      DescriptionToLong = 3;
      IdDontExists = 4;
      DbAccesError = 100;
    }

  repeated Error code = 1;
}

message PositionOfItemOnShelf {

}

message InventoryShelf {
    optional int32 id=1;
    optional string name = 2;
    optional int32 inventoryid = 3;
    optional string positionOfShelf = 4;
    optional string description = 5;

    optional Acl acl = 10;
}
